src/utils/s3-helpers.tsx

```diff
@@ -0,0 +1,70 @@
+import * as AWS from "aws-sdk";
+
+const S3_BUCKET = "models-resources";
+const S3_DIR_PREFIX = "glossary-resources";
+const S3_REGION = "us-east-1";
+const CLOUDFRONT_URL = "https://models-resources.concord.org";
+const CLOUDFRONT_DISTRIBUTION = "E1QHTGVGYD1DWZ";
+
+interface IParams {
+  dir: string;
+  filename: string;
+  accessKey: string;
+  secretKey: string;
+  body: AWS.S3.Types.Body;
+  contentType?: string;
+}
+
+export function s3Upload({ dir, filename, accessKey, secretKey, body, contentType = "" }: IParams) {
+  return new Promise((resolve, reject) => {
+    const s3 = new AWS.S3({
+      region: S3_REGION,
+      accessKeyId: accessKey,
+      secretAccessKey: secretKey,
+    });
+    const key = `${S3_DIR_PREFIX}/${dir}/${filename}`;
+    s3.upload({
+      Bucket: S3_BUCKET,
+      Key: key,
+      Body: body,
+      ACL: "public-read",
+      ContentType: contentType
+    }, (err: Error, data: AWS.S3.ManagedUpload.SendData) => {
+      if (err) {
+        reject(err.message);
+      } else {
+        // Construct custom Cloudfront URL instead of the direct S3 URL.
+        resolve(`${CLOUDFRONT_URL}/${data.Key}`);
+      }
+    });
+    // Invalidate Cloudfront path too.
```

Comment:
I personally would avoid cloudfront invalidations. They provide a false sense of fixing caching problems.  If you are are running into caching problems and you need to invalidate cloudfront, it would be better decrease the caching of the file. I'd guess for the json files it could be set to 0 seconds. And really long for the image and videos assuming they always have unique files names and you aren't updating them in place.

Reply:
I looked at this more. And yes it would be better to set the caching behavior so it is short or 0 for the json.  Otherwise authors that don't clear their browser cache will be getting old versions of the glossary and not know why.

And I see the media files get uuid's prepended, so their caching time could be set to 1 year.

Reply:
I've used the existing Cloudfront distribution (models-resources), so I can't really change its config I guess, as it would affect many projects.

The main problem is that if you save your changes and then load this JSON again, you won't see changes for a while. So, I think we shouldn't use caching for JSON, at least during authoring. Invalidations seemed to work fine for me, there was a tiny delay (~2 secons). But right, the browser cache could be more problematic (I had dev tools open when I was playing with it).

I can create a new distribution for the directory I'm using here and set custom caching rules or simply return direct S3 links to JSON. Direct S3 links seem simpler and less likely to get broken (no config in AWS involved). I'm wondering if there's any benefit of using Cloudfront with 0 caching time vs a direct S3 link? Cloudfront should hit S3 each time anyway. So, I'd rather use S3 links for JSONs and Cloudfront links for media.

Reply:
I might be reading it wrong, but I think the models-resources cloudfront distribution will always prefer the cache headers sent by the origin.  So when the files are uploaded using s3.upload you can set the cache headers. You can see it listed as the CacheControl parameter here: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
That header saved in S3 will override any settings in CloudFormation, and it should also be sent down to the user's browser.

I'm not sure if we get a benefit by using cloudfront with 0 caching. At least it would be uniform so all requests would be going through our concord.org domain. That might help with some schools that filter s3 domains. Actually one thing we get is automatic compression. If configured correctly the cloudfront distribution will gzip all json. I don't think S3 will do that by itself.

Reply:
Just to be clear for other use cases. Even if the CF distribution customizes the Object Caching, it will not send this down to the browser. So using that to configure caching is not an alternative to setting the caching in S3.

Reply:
I checked and the models-resources distribution is configured to compress objects automatically.